contracts/Staking.sol

```diff
@@ -39,17 +65,107 @@ contract Staking is AccessControlEnumerable {
         isWhitelistEnabled = false;
     }
 
-    function isWhitelisted(address caller)
-        external
-        payable
-        onlyRole(TRANSPARENT_FORWARDER_ROLE)
-        returns (bool)
-    {
-        return isWhitelistEnabled ? hasRole(WHITELISTED_ROLE, caller) : true;
+    function pause() external onlyRole(PAUSE_ROLE) {
+        Pausable._pause();
+    }
+
+    function unpause() external onlyRole(PAUSE_ROLE) {
+        Pausable._unpause();
+    }
+
+    function disableEscapeHatch() external onlyRole(DEFAULT_ADMIN_ROLE) {
+        escapeHatchEnabled = false;
+    }
+
+    function isWhitelisted(
+        address caller
+    ) external payable onlyRole(TRANSPARENT_FORWARDER_ROLE) returns (bool) {
+        return isWhitelistEnabled ? clientStake[caller] >= minStake : true;
     }
 
     function withdraw() external onlyRole(ESCAPE_HATCH_ROLE) {
         uint256 amount = address(this).balance;
         payable(msg.sender).transfer(amount);
     }
+
+    function escape(
+        address _address
+    ) external onlyRole(ESCAPE_HATCH_ROLE) whenPaused {
+        if (escapeHatchEnabled) {
+            require(
+                token.transfer(_address, token.balanceOf(address(this))),
+                "token transfer failed"
+            );
+        } else {
+            revert("escape hatch is disabled");
+        }
+    }
+
+    function stake(address client, uint256 amount) external whenNotPaused {
+        require(amount > 0, "amount must be greater than 0");
```

Comment:
let's set a `minsafestake` which can be less than minstake. Here we are enabling any user to stake as long as they have even 1 wei of razor, which may allow few attack vectors in the future